### PR TITLE
PT-3045: Facets for multivalue, multilanguage properties

### PR DESCRIPTION
**Pull request by akak1977**
## Description
Proper facets handling for multivalue, multilanguage properties.
### Jira-link:
https://virtocommerce.atlassian.net/browse/PT-3044

### CODE REVIEW

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogBrowseFiltersController.cs [hunk 1]
```diff
@@ -91,10 +91,35 @@ public async Task<ActionResult> SetAggregationProperties(string storeId, [FromBo
...
+                    catalogProperty.Multivalue &&
+                    catalogProperty.Multilanguage)
+                {
+                    foreach (var lang in store.Languages)
+                    {
+                        var aggregationPropertyLangSpecific=aggregationProperty.Clone() as AggregationProperty;
+                        aggregationPropertyLangSpecific.Name = $"{aggregationPropertyLangSpecific.Name}_{lang.ToLowerInvariant()}";
+                        browseFilterPropertiesList.Add(aggregationPropertyLangSpecific);
+                    }
+                }
```
Comment (Vectorfield4) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogBrowseFiltersController.cs:117]:
```suggestion
                if (catalogProperty is null || catalogProperty.Dictionary || !catalogProperty.Multivalue || !catalogProperty.Multilanguage)
                {
                    continue;
                }

                // Comment here what going on
                foreach (var cultureName in store.Languages)
                {
                    var aggregationPropertyLangSpecific=aggregationProperty.Clone() as AggregationProperty;
                    aggregationPropertyLangSpecific.Name = $"{aggregationPropertyLangSpecific.Name}_{cultureName.ToLowerInvariant()}";
                    browseFilterPropertiesList.Add(aggregationPropertyLangSpecific);
                }
```

**Review by Vectorfield4 — APPROVED** [review]

**Review by ksavosteev — APPROVED** [review]

File: src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogBrowseFiltersController.cs [hunk 2]
```diff
@@ -91,10 +91,38 @@ public async Task<ActionResult> SetAggregationProperties(string storeId, [FromBo
...
+
+            var catalogProperties = await _propertyService.GetAllCatalogPropertiesAsync(store.Catalog);
+
+            var browseFilterPropertiesList = new List<AggregationProperty>();
+            foreach (var aggregationProperty in browseFilterProperties)
+            {
+                browseFilterPropertiesList.Add(aggregationProperty);
+                var catalogProperty = catalogProperties.FirstOrDefault(x => x.Name == aggregationProperty.Name);
+                // If the property is mulivalue, multilanguage, but not dictionary, let's add synthetic aggregation property for each store culture
+                // To allow future facet filtering.
```
Comment (tatarincev) [inline src/VirtoCommerce.CatalogModule.Web/Controllers/Api/CatalogBrowseFiltersController.cs:107]: Why we do not add synthetic facets for single value Multilang properties?

File: src/VirtoCommerce.CatalogModule.Web/Model/AggregationProperty.cs [hunk 3]
```diff
@@ -1,15 +1,21 @@
...
         public string Type { get; set; }
         public string Currency { get; set; }
         public int? Size { get; set; }
         public int ValuesCount => Values?.Count ?? 0;
-        public IList<string> Values { get; set; }
+        public IList<string> Values { get; set; }
+
+        public object Clone()
+        {
+            return MemberwiseClone();
```
Comment (tatarincev) [inline src/VirtoCommerce.CatalogModule.Web/Model/AggregationProperty.cs:18]: Values list also must be cloned 
`result.Values = Values?.ToList()`

**Review by tatarincev — COMMENTED** [review]
These changes introduces a  side-effect for existing clients who search products with this enpoint
​POST /api​/catalog​/search​/products
since it now will return "synthetic" facets along with normal propery
image

But I think we should do nothing with this for a while.

**Pull request merged**